src/github.com/matrix-org/dendrite/clientapi/readers/profile.go

```diff
@@ -175,13 +238,96 @@ func SetDisplayName(
 	}
 }
 
-func getLocalPart(userID string) string {
-	if !strings.HasPrefix(userID, "@") {
-		panic(fmt.Errorf("Invalid user ID"))
+func buildMembershipEvents(
+	memberships []authtypes.Membership, db *accounts.Database,
+	oldProfile *authtypes.Profile, changedKey string, newValue string,
+	userID string, cfg *config.Dendrite, queryAPI api.RoomserverQueryAPI,
+) ([]gomatrixserverlib.Event, error) {
+	ev := []gomatrixserverlib.Event{}
+
+	for _, membership := range memberships {
+		prevContent := events.MemberContent{
+			Membership:  "join",
+			DisplayName: oldProfile.DisplayName,
+			AvatarURL:   oldProfile.AvatarURL,
+		}
+
+		prev := prevMembership{
+			UserID:      userID,
+			PrevID:      membership.EventID,
+			PrevContent: prevContent,
+		}
+
+		builder := gomatrixserverlib.EventBuilder{
+			Sender:   userID,
+			RoomID:   membership.RoomID,
+			Type:     "m.room.member",
+			StateKey: &userID,
+		}
+
+		if err := builder.SetUnsigned(prev); err != nil {
+			return nil, err
+		}
+
+		content := events.MemberContent{
+			Membership: "join",
+		}
+
+		if changedKey == "displayname" {
+			content.DisplayName = newValue
+			content.AvatarURL = oldProfile.AvatarURL
+		} else if changedKey == "avatar_url" {
+			content.DisplayName = oldProfile.DisplayName
+			content.AvatarURL = newValue
+		}
+
+		if err := builder.SetContent(content); err != nil {
+			return nil, err
+		}
+
+		eventsNeeded, err := gomatrixserverlib.StateNeededForEventBuilder(&builder)
+		if err != nil {
+			return nil, err
+		}
+
+		// Ask the roomserver for information about this room
+		queryReq := api.QueryLatestEventsAndStateRequest{
+			RoomID:       membership.RoomID,
+			StateToFetch: eventsNeeded.Tuples(),
+		}
+		var queryRes api.QueryLatestEventsAndStateResponse
+		if queryErr := queryAPI.QueryLatestEventsAndState(&queryReq, &queryRes); queryErr != nil {
+			return nil, err
+		}
+
+		authEvents := gomatrixserverlib.NewAuthEvents(nil)
+
+		// Iterating the old way because range seems to mess things up. Might be
+		// worth investigating.
+		for i := 0; i < len(queryRes.StateEvents); i++ {
+			authEvents.AddEvent(&queryRes.StateEvents[i])
+		}
+
+		refs, err := eventsNeeded.AuthEventReferences(&authEvents)
+		if err != nil {
+			return nil, err
+		}
+		builder.AuthEvents = refs
+
+		// Generate a new event ID and set it in the database
+		eventID := fmt.Sprintf("$%s:%s", util.RandomString(16), cfg.Matrix.ServerName)
+		if err = db.UpdateMembership(membership.EventID, eventID); err != nil {
+			return nil, err
+		}
```

Comment:
We shouldn't be updating the membership here, we should only be updating the membership using the output from the roomserver.

The idea is that the roomserver can act as a single source of truth for the state of the room. If  we update the state here then we could end up in with a view of the membership that is inconsistent with the room server thinks the state is.